src/commonMain/kotlin/fr/acinq/lightning/channel/Channel.kt

```diff
@@ -283,6 +234,129 @@ sealed class ChannelState {
         else -> actions
     }
 
+    /** Add actions for some transitions */
+    private fun ChannelContext.onTransition(newState: ChannelState): List<ChannelAction> {
+        val oldState = when (this@ChannelState) {
+            is Offline -> this@ChannelState.state
+            is Syncing -> this@ChannelState.state
+            else -> this@ChannelState
+        }
+        return when {
+            // we only want to fire the PaymentSent event when we transition to Closing for the first time
+            oldState is ChannelStateWithCommitments && oldState !is WaitForInit && oldState !is Closing && newState is Closing -> emitClosingEvents(oldState, newState)
+            else -> emptyList()
+        }
+    }
+
+    private fun ChannelContext.emitClosingEvents(oldState: ChannelStateWithCommitments, newState: Closing): List<ChannelAction> {
+        val channelBalance = oldState.commitments.latest.localCommit.spec.toLocal
```

Comment:
How can the mining fee be negative, since `channelBalance` is the balance before any mining fees, and we subtract the output of 1st, 2nd and 3rd level transactions?

Reply:
If for example we have 5_000 sat in our main output and 10_000 sat in an HTLC, that will yield:

- channelBalance = 5_000 sat
- finalAmount = slightly less than 15_000 sat

So the mining fee will be -10_000 sat...

Reply:
Gotcha! Removed the buggy estimate in [8c452f0](https://github.com/ACINQ/lightning-kmp/pull/436/commits/8c452f01e41d8fa83ec75866e50e40fa31565f59)